File: src/main/java/com/lj/datastructure/notline/UnionFind.java
package com.lj.datastructure.notline;

/**
 * 并查集
 */
public class UnionFind {

    public int[] root;
    public int cnt;

    public UnionFind(int size) {
        root = new int[size];
        cnt = size;
        for (int i = 0; i < size; i++) {
            root[i] = i;
        }
    }

    public void merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) {
            root[x] = y;
            cnt --;
        }
    }

    private int find(int i) {
        if (root[i] == i) {
            return i;
        }
        // 路径压缩
        return root[i] = find(root[i]);
    }

}

File: src/main/python/datastructure/UnionFind.py
# 并查集

class UnionFind:
    def __init__(self, size: int):
        self.root = [i for i in range(size)]

    def find(self, x: int):
        if x != self.root[x]:
            self.root[x] = self.find(self.root[x])
        return self.root[x]

    def merger(self, x: int, y: int) -> bool:
        x = self.find(x)
        y = self.find(y)
        if x != y:
            self.root[x] = y
            return True
        return False


